File: src/ShortestDistance.java
import java.lang.Math;
public class ShortestDistance{
    int countOfEuc = 0;

    public double euclideanDistance(point p1, point p2){
        countOfEuc++;
        double buffer = 0;
        for(int i = 0; i < p1.derajat; i++){
            buffer += (p1.buffer[i]-p2.buffer[i]) * (p1.buffer[i]-p2.buffer[i]);
        }
        return Math.sqrt(buffer);
    }

    public ReturnType findShortestDistance(ArrayPoint ap){
        // ap sudah terurut terhadap x
        ReturnType ret;
        if(ap.n == 2){
            ret = new ReturnType(ap.array[0], ap.array[1], euclideanDistance(ap.array[0], ap.array[1]));
            return ret;
        }else if(ap.n == 3){
            double t1 = euclideanDistance(ap.array[0], ap.array[1]);
            double t2 = euclideanDistance(ap.array[0], ap.array[2]);
            double t3 = euclideanDistance(ap.array[1], ap.array[2]);
            if(t1 < t2){
                if(t1 < t3){
                    ret = new ReturnType(ap.array[0], ap.array[1], t1);
                }else{
                    ret = new ReturnType(ap.array[1], ap.array[2], t3);
                }
            }else{
                if(t2 < t3){
                    ret = new ReturnType(ap.array[0], ap.array[2], t2);
                }else{
                    ret = new ReturnType(ap.array[1], ap.array[2], t3);
                }
            }
            return ret;
        }else{
            // DIVIDE
            // Define midpoint
            int carry;
            if(ap.n % 2 == 0){
                carry = 0;
            }else{
                carry = 1;
            }

            point[] p1_arr = new point[ap.n/2];
            point[] p2_arr = new point[ap.n/2 + carry];

            for(int i = 0; i < ap.n/2; i++){
                p1_arr[i] = ap.array[i];
            }

            for(int i = 0; i < ap.n/2 + carry; i++){
                p2_arr[i] = ap.array[i + ap.n/2];
            }

            ArrayPoint p1 = new ArrayPoint(ap.n/2, p1_arr);
            ArrayPoint p2 = new ArrayPoint(ap.n/2 + carry, p2_arr);
            
            // Combine
            ReturnType d1 = findShortestDistance(p1);
            ReturnType d2 = findShortestDistance(p2);

            // Conquer
            if(d1.dist < d2.dist){
                ret = new ReturnType(d1.getP1(), d1.getP2(), d1.dist);
            }else{
                ret = new ReturnType(d2.getP1(), d2.getP2(), d2.dist);
            }

            // Make Slab
            int nslab = 0;
            for(int i = 0; i < p1.n; i++){
                if(p1.array[i].getX() >= ap.array[ap.n/2].getX()-ret.dist){
                    nslab++;
                }
            }

            for(int i = 0; i < p2.n; i++){
                if(p2.array[i].getX() <= ap.array[ap.n/2].getX()+ret.dist){
                    nslab++;
                }
            }

            ArrayPoint slab = new ArrayPoint(nslab);
            slab.array = new point[nslab];
            // fill the slab
            int nfill = 0;

            for(int i = 0; i < p1.n; i++){
                if(p1.array[i].getX() >= ap.array[ap.n/2].getX()-ret.dist){
                    slab.array[nfill] = p1.array[i];
                    nfill++;
                }
            }

            for(int i = 0; i < p2.n; i++){
                if(p2.array[i].getX() <= ap.array[ap.n/2].getX()+ret.dist){
                    slab.array[nfill] = p2.array[i];
                    nfill++;
                }
            }

            /* Iterasikan untuk jarak terdekat,
             periksa untuk seluruh nilai n, n-1, ..., z, dan y */
            for(int i = ap.array[0].derajat-1; i > 0 ;i--){
                slab.sortArray(slab.array, i);
                for (int j = 0; j < slab.n; j++){
                    for(int k = j+1; k < slab.n; k++){
                        if(Math.abs(slab.array[j].getX() - slab.array[k].getX()) < ret.dist || Math.abs(slab.array[j].buffer[i] - slab.array[k].buffer[i]) < ret.dist){
                            ReturnType newret = new ReturnType(slab.array[j], slab.array[k], euclideanDistance(slab.array[j], slab.array[k]));
                            if(ret.dist > newret.dist){
                                ret = newret;
                            }
                        }
                    }
                }
            }

            return ret;
        }
    }

    public ReturnType findShortestDistanceBrute(ArrayPoint ap){
        // Mencari Shortest Distance Menggunakan Algoritma Bruteforce
        double d = euclideanDistance(ap.array[0], ap.array[1]);
        point p1 = ap.array[0];
        point p2 = ap.array[1];
        for(int i = 0; i < ap.n; i++){
            for(int j = 0; j < ap.n; j++){
                if(i != j){
                    double temp = euclideanDistance(ap.array[i], ap.array[j]);
                    if(temp < d){
                        d = temp;
                        p1 = ap.array[i];
                        p2 = ap.array[j];
                    }
                }
            }
        }

        ReturnType ret = new ReturnType(p1,p2,d);

        return ret;
    }
}

File: src/ShortestDistance.py
# Filename: ShortestDistance.py
# Matthew Mahendra - 13521007
import point as pt
import math

countOfEuc = 0

# Fungsi untuk menentukan euclidean distance antar dua titik
def euclideanDistance(p1, p2):
    global countOfEuc
    countOfEuc += 1
    d = len(p1)
    buffer = 0.0
    for i in range(d):
        buffer+=(p1[i]-p2[i])*(p1[i]-p2[i])
    return math.sqrt(buffer)

# Fungsi untuk menentukan jarak terdekat dari set of points
def findShortestDistance(ap, d):
    # Jumlah titik dengan menghitung panjang dari array of points
    n = len(ap)
    # Basis ketika jumlah titik adalah 2 atau 3
    if(n == 2):
        return [ap[0],ap[1], euclideanDistance(ap[0], ap[1])]
    elif (n==3):
        t1 = euclideanDistance(ap[0], ap[1])
        t2 = euclideanDistance(ap[0], ap[2])
        t3 = euclideanDistance(ap[1], ap[2])

        if(t1 < t2):
            if(t1 < t3):
                ret = [ap[0], ap[1], t1]
            else:
                ret = [ap[1], ap[2], t3]
            
        else:
            if(t2 < t3):
                ret = [ap[0], ap[2], t2]
            else:
                ret = [ap[1], ap[2], t3]
        
        return ret
    else:
        # Divide
        if(n % 2 == 0):
            carry = 0
        else:
            carry = 1
        
        # Bagi titik-titik menjadi partisi p1 dan p2
        p1 = [[] for i in range (n//2)]
        p2 = [[] for i in range ((n//2)+carry)]
        for i in range(n//2):
            p1[i] = ap[i]

        for i in range((n//2) + carry):
            p2[i] = ap[i + (n//2)]

        # Cari jarak minimum pada tiap partisi
        d1 = findShortestDistance(p1, d)
        d2 = findShortestDistance(p2, d)

        # Conquer
        # Bandingkan d1 dan d2 untuk mengambil minimumnya
        if(d1[2] < d2[2]):
            ret = [d1[0], d1[1], d1[2]]
        else:
            ret = [d2[0], d2[1], d2[2]]

        # Combine
        # Make a Slab

        # Hitung jumlah yang akan dimasukkan ke dalam slab
        nslab = 0
        for i in range(len(p1)):
            if(p1[i][0] >= ap[n//2][0]-ret[2]):
                nslab += 1
        
        for i in range(len(p2)):
            if(p2[i][0] <= ap[n//2][0]+ret[2]):
                nslab += 1

        slab = [[] for i in range(nslab)]

        nfill = 0
        for i in range(len(p1)):
            if(p1[i][0] >= ap[n//2][0]-ret[2]):
                slab[nfill] = p1[i]
                nfill += 1

        for i in range(len(p2)):
            if(p2[i][0] <= ap[n//2][0]+ret[2]):
                slab[nfill] = p2[i]
                nfill += 1
        
        for i in range(d-1, 0,-1):
            slab = pt.sortArray(slab, i)
            for j in range(len(slab)):
                for k in range(j+1, len(slab)):
                    if(abs(slab[j][0] - slab[k][0]) < ret[2] or abs(slab[j][i] - slab[k][i]) < ret[2]):
                        newret = [slab[j], slab[k], euclideanDistance(slab[j], slab[k])]
                        if(ret[2] > newret[2]):
                            ret = newret
        
        return ret

def findShortestDistanceBrute(ap):
    # Mencari Shortest Distance Menggunakan Algoritma Bruteforce
    d = euclideanDistance(ap[0], ap[1])
    p1 = ap[0]
    p2 = ap[1]
    for i in range(len(ap)):
        for j in range(i+1, len(ap)):
            temp = euclideanDistance(ap[i], ap[j])
            if(temp < d):
                d = temp
                p1 = ap[i]
                p2 = ap[j]

    ret = [p1,p2,d]

    return ret